File: 6주차/LeetCode_22_Generate_Parentheses/dongjun.java
/**
 * @문제접근
 * 백트래킹을 통해 풀었습니다.
 * 괄호가 성립하는 조건에 맞춰 탐색을 진행했습니다.
 *
 * @성능
 * Runtime: 1 ms
 * Memory Usage: 39.4 MB
 */
public class GenerateParentheses {
    public static ArrayList<String> answer = new ArrayList<>();

    public List<String> generateParenthesis(int n) {
        answer.clear();
        dfs("(", 1, 0, n);

        return answer;
    }

    public void dfs(String now, int lbCnt, int rbCnt, int n){
        if(now.length() == n * 2){
            answer.add(now);
            return;
        }
        if(lbCnt < n)
            dfs(new String(now + "("), lbCnt + 1, rbCnt, n);

        if(lbCnt <= n && lbCnt > rbCnt)
            dfs(new String(now + ")"), lbCnt, rbCnt + 1, n);
    }
}

File: 11주차/BOJ_13023_ABCDE/dongjun.py
'''
1. 문제접근
    dfs로 풀었습니다.

2. 성능
    Runtime: 1280ms
    Memory Usage: 29380KB
    시간복잡도: O(N^3)?
'''
import sys
input = sys.stdin.readline


def main():
    N, M = map(int, input().split())

    relationship = [tuple(map(int, input().split())) for _ in range(M)]

    print(solve(N, M, relationship))


def solve(n, m, relationship):
    adj_list = [[] for _ in range(n)]
    visited = [False] * n

    for a, b in relationship:
        adj_list[a].append(b)
        adj_list[b].append(a)

    def dfs(now, cnt):
        nonlocal adj_list, visited
        if cnt == 4:
            print(1)
            return exit(0)

        visited[now] = True
        friends = adj_list[now]

        for friend in friends:
            if not visited[friend]:
                dfs(friend, cnt + 1)

        visited[now] = False

    for person in range(n):
        dfs(person, 0)

    return 0


if __name__ == '__main__':
    main()
